core/src/main/java/org/jdbi/v3/core/LazyHandleSupplier.java

```diff
@@ -52,45 +56,56 @@ public Jdbi getJdbi() {
 
     @Override
     public Handle getHandle() {
-        if (handle == null) {
-            initHandle();
-        }
-        return handle;
+        return handleHolder.get();
     }
 
-    private void initHandle() {
+    private Handle createHandle() {
+        if (closed.get()) {
+            throw new IllegalStateException("Handle is closed");
+        }
         synchronized (lock) {
-            if (handle == null) {
-                if (closed) {
-                    throw new IllegalStateException("Handle is closed");
-                }
-
-                handle = db.open();
-                // share extension method thread local with handle,
-                // so extension methods set in other threads are preserved
-                handle.setExtensionMethodThreadLocal(localExtensionMethod);
-                handle.setLocalConfig(localConfig);
-            }
+            return db.open().acceptExtensionContext(extensionContexts.peek());
         }
     }
 
     @Override
     public <V> V invokeInContext(ExtensionMethod extensionMethod, ConfigRegistry config, Callable<V> task) throws Exception {
-        return invokeWith(localExtensionMethod, extensionMethod,
-                () -> invokeWith(localConfig, config, task));
+        return invokeInContext(new ExtensionContext(config, extensionMethod), task);
+    }
+
+    @Override
+    public <V> V invokeInContext(ExtensionContext extensionContext, Callable<V> task) throws Exception {
+        try {
+            pushExtensionContext(extensionContext);
+            return task.call();
+        } finally {
+            popExtensionContext();
+        }
+    }
+
+    private void pushExtensionContext(ExtensionContext extensionContext) {
+        synchronized (lock) {
+            extensionContexts.addFirst(extensionContext);
+            handleHolder.ifInitialized(h -> h.acceptExtensionContext(extensionContext));
+        }
+    }
+
+    private void popExtensionContext() {
+        synchronized (lock) {
+            extensionContexts.pollFirst();
+            handleHolder.ifInitialized(h -> h.acceptExtensionContext(extensionContexts.peek()));
+        }
     }
 
     @Override
     public void close() {
+        if (closed.get()) {
+            throw new IllegalStateException("Handle is closed");
+        }
+
         synchronized (lock) {
-            closed = true;
```

Comment:
I don't see a corresponding `closed.set(true)` to replace the old `closed = true`